libopae/plugins/xfpga/metrics/metrics_utils.c

```diff
@@ -0,0 +1,1231 @@
+			}
+
+		}
+
+		if (strcmp(dirent->d_name, PERF_FABRIC) == 0) {
+			snprintf_s_ss(qualifier_name, sizeof(qualifier_name), "%s:%s", PERF, PERF_FABRIC);
+			result = enum_perf_counter_items(vector, metric_num, qualifier_name, sysfs_path, dirent->d_name, FPGA_METRIC_TYPE_PERF, hw_type);
+			if (result != FPGA_OK) {
+				FPGA_MSG("Failed to add metrics");
+			}
+
+		}
+
+		if (strcmp(dirent->d_name, PERF_IOMMU) == 0) {
+			snprintf_s_ss(qualifier_name, sizeof(qualifier_name), "%s:%s", PERF, PERF_IOMMU);
+			result = enum_perf_counter_items(vector, metric_num, qualifier_name, sysfs_path, dirent->d_name, FPGA_METRIC_TYPE_PERF, hw_type);
+			if (result != FPGA_OK) {
+				FPGA_MSG("Failed to add metrics");
+			}
+
+		}
+
+	}
+
+	return result;
+}
+
+fpga_result xfpga_bmcLoadSDRs(struct _fpga_handle *_handle, bmc_sdr_handle *records,
+	uint32_t *num_sensors)
+{
+	fpga_result result = FPGA_NOT_FOUND;
+	fpga_result(*bmcLoadSDRs)(fpga_token token, bmc_sdr_handle *records,
+		uint32_t *num_sensors);
+	if (_handle->dl_handle != NULL) {
+		bmcLoadSDRs = dlsym(_handle->dl_handle, "bmcLoadSDRs");
+
+		result = bmcLoadSDRs(_handle->token, records, num_sensors);
+
+	}
+	return result;
+}
+
+fpga_result xfpga_bmcDestroySDRs(struct _fpga_handle *_handle, bmc_sdr_handle *records)
+{
+	fpga_result result = FPGA_NOT_FOUND;
+	fpga_result(*bmcDestroySDRs)(bmc_sdr_handle *records);
+
+	if (_handle->dl_handle != NULL) {
+
+		bmcDestroySDRs = dlsym(_handle->dl_handle, "bmcDestroySDRs");
+
+		result = bmcDestroySDRs(records);
+
+	}
+	return result;
+}
+
+
+fpga_result xfpga_bmcReadSensorValues(struct _fpga_handle *_handle, bmc_sdr_handle records,
+	bmc_values_handle *values,
+	uint32_t *num_values)
+{
+	fpga_result result = FPGA_NOT_FOUND;
+	fpga_result(*bmcReadSensorValues)(bmc_sdr_handle records, bmc_values_handle *values, uint32_t *num_values);
+
+	if (_handle->dl_handle != NULL) {
+
+		bmcReadSensorValues = dlsym(_handle->dl_handle, "bmcReadSensorValues");
+
+		result = bmcReadSensorValues(records, values, num_values);
+
+	}
+	return result;
+}
+
+
+fpga_result xfpga_bmcDestroySensorValues(struct _fpga_handle *_handle, bmc_values_handle *values)
+{
+	fpga_result result = FPGA_NOT_FOUND;
+	fpga_result(*bmcDestroySensorValues)(bmc_values_handle *values);
+
+	if (_handle->dl_handle != NULL) {
+
+		bmcDestroySensorValues = dlsym(_handle->dl_handle, "bmcDestroySensorValues");
+
+		result = bmcDestroySensorValues(values);
+
+	}
+	return result;
+}
+
+fpga_result xfpga_bmcGetSensorReading(struct _fpga_handle *_handle, bmc_values_handle values,
+	uint32_t sensor_number, uint32_t *is_valid,
+	double *value)
+{
+	fpga_result result = FPGA_NOT_FOUND;
+	fpga_result(*bmcGetSensorReading)(bmc_values_handle values,
+		uint32_t sensor_number, uint32_t *is_valid,
+		double *value);
+
+	if (_handle->dl_handle != NULL) {
+
+		bmcGetSensorReading = dlsym(_handle->dl_handle, "bmcGetSensorReading");
+
+		result = bmcGetSensorReading(values, sensor_number, is_valid, value);
```

Comment:
Need to check whether bmcGetSensorReading was resolved properly before calling it.